File: Dynamic Programming/Java/Unique Binary Search Trees.java
// Problem link: https://leetcode.com/problems/unique-binary-search-trees/description/
// Time: O(n ^ 2)
// Space: O(n)
class Solution {
    public int numTrees(int n) {
        if(n < 3) {
            return n;
        }

        return this.DP(n, new int[n + 1]);
    }

    private int DP(int n, int[] cache) {
        if(n <= 1) {
            return 1;
        }
        else if(cache[n] != 0) {
            return cache[n];
        }

        int result = 0;

        for(int num = 1; num <= n; num++) {
            result += this.DP(num - 1, cache) * this.DP(n - num, cache);
        }

        cache[n] = result;
        return cache[n];
    }
}

File: Dynamic Programming/Python/Unique Binary Search Trees.py
# Problem link: https://leetcode.com/problems/unique-binary-search-trees/description/
# Time: O(n ^ 2)
# Space: O(n)
class Solution:
    def numTrees(self, n: int) -> int:
        if n < 3:
            return n
        
        return self.DP(n, [0] * (n + 1))
    
    def DP(self, n, cache):
        if n <= 1:
            return 1
        elif cache[n] != 0:
            return cache[n]
        
        result = 0

        for i in range(1, n + 1):
            result += self.DP(i - 1, cache) * self.DP(n - i, cache)
        
        cache[n] = result
        return result
